accepted/0000-hub_retail.md

```diff
@@ -0,0 +1,213 @@
+  - Branch PXE data - entries describing how to boot each terminal
+  
+# Detailed design
+[design]: #detailed-design
+
+Images + Image metadata
+-----------------------
+
+Current situation: an image is built on Buildhost, then pushed to SUMA Server htdocs directory. Then SUMA creates a db entry (including package list) and a pillar file.
+The pillar conains URL of image tarball, checksums and additional deployment details. It is used for image synchronization and deployment. These data are extracted 
+on Buildhost at the end of build phase. Some of the data can't be extracted later form the tarball.
+
+The SUMA API related to images is rather docker-centered. For Kiwi images it allows to list images and packages and delete an image.
+
+Use cases currently not covered:
+
+- Move image tarball to external HTTP server
+  - In large environments it makes sense to use one or more servers for distribution of images, it should not be the Hub itself.
+- Copy image metadata from one SUMA Server to another 
+  - Hub - copy to all servers
+  - Test in lab, then move the same image to production.
+- Adjust deployment details
+
+
+Proposed solution:
+
+- Implement API calls to
+  - Get Kiwi Image pillar
+  - Update Kiwi Image pillar
+  - Export Kiwi Image (complete metadata)
+  - Import Kiwi Image
+
+- Implement a tool that handles the use cases above
+
+
+HWTYPE configurations
+---------------------
+
+This is currently implemented as a formula assigned to a group with special name.
+
+FIXME: pick one alternative
+
+Alternative 1:
+Extend retail_yaml to create the groups and formula data via Hub and Peripheral Server API.
+This is relatively small change to the code.
+
+Alternative 2:
+Central creation of groups is covered in [Hub Users RFC](00061-hub-users.md). The salt states
+can be extended to handle also formula data.
+This approach is probably more flexible.
+
+Both alternatives will have to cope with authentization and Hub vs. Peripheral server user credentials.
+
+YAML configuration
+------------------
+
+FIXME: optional, related to Alternative 2 in previous section
+
+We have retail_yaml tool for initial configuration of Retail environment from YAML file.
+[Hub Users RFC](00061-hub-users.md) introduced similar concept, based on Salt.
+Maybe these parts could be merged.
+
+
+Combined SUMA and Branch server
+-------------------------------
+
+Branch services are configured by Salt formulas. Also PXE entries for each terminal are generated
+by Salt (currently via reactor).
+With combined SUMA and Branch server, these formulas should be applied on SUMA server itself.
+
+Terminals connect to SUMA directly, SUMA Proxy is not needed.
+
+
+FIXME: pick one alternative
+
+Alternative 1: Hub as a Salt Master for Combined SUMA and Branch server
+
+- this is already used for [Hub Users RFC](00061-hub-users.md)
+
+problems:
+- does not work for PXE entries and disconnected operation
```

Comment:
I am not sure why it would not work for disconnected scenarios? There will need to be an access to hub when configuring services (downloading image, applying states). But after that it should work disconnected.

What is here to solve is pxe generation as for that we need to be able to apply states.

Reply:
I mean that disconnected scenarios do not work with PXE entries and maybe in the future for generating static DNS and DHCP entries.
Yes, the event must be terminated on peripheral server,  otherwise we would have the same scalability problem as before.

Reply:
Where are PXE entries supposed to be set and when, and by what component?

Reply:
During saltboot process terminal sends event `pxe_update`, this event is processed by reactor and applies `pxe/terminal_entry` state on branch server.

This happens after terminal is deployed, but before switching from initrd to system image. In case kernel of initrd does not match kernel of system, terminal is tried to kexec to new kernel. However if kexec is not possible or fails, terminal is rebooted. For this pxe entry needs to be already prepared as it specifies correct kernel and initrd to use for the machine.